satellite/gracefulexit/endpoint.go

```diff
@@ -509,71 +408,55 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
 	if err != nil {
 		return Error.Wrap(err)
 	}
-	remote := pointer.GetRemote()
-
-	pieces := remote.GetRemotePieces()
-	var nodePiece *pb.RemotePiece
-	excludedNodeIDs := make([]storj.NodeID, len(pieces))
-	for i, piece := range pieces {
-		if piece.NodeId == nodeID && piece.PieceNum == incomplete.PieceNum {
-			nodePiece = piece
-		}
-		excludedNodeIDs[i] = piece.NodeId
-	}
-
-	if nodePiece == nil {
-		endpoint.log.Debug("piece no longer held by node", zap.Stringer("Node ID", nodeID), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
 
-		err = endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path, incomplete.PieceNum)
+	// validate pointer state
+	nodePiece, err := validatePointer(ctx, pointer, incomplete)
+	if err != nil {
+		endpoint.log.Debug("piece no longer held by node", zap.Stringer("node ID", nodeID), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
+		err := endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path, incomplete.PieceNum)
 		if err != nil {
 			return Error.Wrap(err)
 		}
+		return Error.Wrap(err)
+	}
 
+	pieceSize, err := endpoint.calculatePieceSize(ctx, pointer, incomplete, nodePiece)
+	if ErrAboveOptimalThreshold.Has(err) {
```

Comment:
To me, it would make more sense for `calculatePieceSize` to not do any logic related to updating metainfo or deleting from the transfer queue. Rather, it should just return the `ErrAboveOptimalThreshold` and we should do those updates here or in a separate function called from here.